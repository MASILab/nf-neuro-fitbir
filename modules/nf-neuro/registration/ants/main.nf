
process REGISTRATION_ANTS {
    tag "$meta.id"
    label 'process_high'

    container "scilus/scilus:2.2.0"

    input:
    tuple val(meta), path(fixedimage), path(movingimage), path(mask) /* optional, input = [] */

    output:
    tuple val(meta), path("*_warped.nii.gz")                        , emit: image
    tuple val(meta), path("*__output0GenericAffine.mat")             , emit: affine
    tuple val(meta), path("*__output1InverseAffine.mat")            , emit: inverse_affine
    tuple val(meta), path("*__output1Warp.nii.gz")                  , emit: warp, optional:true
    tuple val(meta), path("*__output0InverseWarp.nii.gz")           , emit: inverse_warp, optional: true
    tuple val(meta), path("*_registration_ants_mqc.gif")            , emit: mqc, optional: true
    path "versions.yml"                                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def initialization_types = ["geometric center": 0, "intensities": 1, "origin": 2]
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix_qc = task.ext.suffix_qc ? "${task.ext.suffix_qc}" : ""
    def ants = task.ext.quick ? "antsRegistrationSyNQuick.sh " :  "antsRegistrationSyN.sh "
    def dimension = task.ext.dimension ? "-d " + task.ext.dimension : "-d 3"
    def transform = task.ext.transform ? task.ext.transform : "s"
    def seed = task.ext.random_seed ? " -e " + task.ext.random_seed : "-e 1234"
    def run_qc = task.ext.run_qc ? task.ext.run_qc : false

    if ( task.ext.threads ) args += "-n " + task.ext.threads
    if ( task.ext.initial_transform ) args += " -i [$fixedimage,$movingimage,${initialization_types[task.ext.initial_transform]}]"
    if ( task.ext.histogram_bins ) args += " -r " + task.ext.histogram_bins
    if ( task.ext.spline_distance ) args += " -s " + task.ext.spline_distance
    if ( task.ext.gradient_step ) args += " -g " + task.ext.gradient_step
    if ( task.ext.mask ) args += " -x $mask"
    if ( task.ext.type ) args += " -p " + task.ext.type
    if ( task.ext.histogram_matching ) args += " -j " + task.ext.histogram_matching
    if ( task.ext.repro_mode ) args += " -y " + task.ext.repro_mode
    if ( task.ext.collapse_output ) args += " -z " + task.ext.collapse_output

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    $ants $dimension -f $fixedimage -m $movingimage -o output -t $transform $args $seed

    mv outputWarped.nii.gz ${prefix}__warped.nii.gz
    mv output0GenericAffine.mat ${prefix}__output0GenericAffine.mat

    if [ $transform != "t" ] && [ $transform != "r" ] && [ $transform != "a" ];
    then
        mv output1InverseWarp.nii.gz ${prefix}__output0InverseWarp.nii.gz
        mv output1Warp.nii.gz ${prefix}__output1Warp.nii.gz
    fi

    antsApplyTransforms -d 3 -i $fixedimage -r $movingimage \
                        -o Linear[${prefix}__output1InverseAffine.mat] \
                        -t [${prefix}__output0GenericAffine.mat,1]

    ### ** QC ** ###
    if $run_qc;
    then
        mv $fixedimage fixedimage.nii.gz
        extract_dim=\$(mrinfo fixedimage.nii.gz -size)
        read sagittal_dim coronal_dim axial_dim <<< "\${extract_dim}"

        # Get the middle slice
        coronal_dim=\$((\$coronal_dim / 2))
        axial_dim=\$((\$axial_dim / 2))
        sagittal_dim=\$((\$sagittal_dim / 2))

        # Set viz params.
        viz_params="--display_slice_number --display_lr --size 256 256"
        # Iterate over images.
        for image in fixedimage warped;
        do
            mrconvert *\${image}.nii.gz *\${image}_viz.nii.gz -stride -1,2,3
            scil_viz_volume_screenshot *\${image}_viz.nii.gz \${image}_coronal.png \
                --slices \$coronal_dim --axis coronal \$viz_params
            scil_viz_volume_screenshot *\${image}_viz.nii.gz \${image}_sagittal.png \
                --slices \$sagittal_dim --axis sagittal \$viz_params
            scil_viz_volume_screenshot *\${image}_viz.nii.gz \${image}_axial.png \
                --slices \$axial_dim --axis axial \$viz_params
            if [ \$image != fixedimage ];
            then
                title="T1 Warped"
            else
                title="fixedimage"
            fi
            convert +append \${image}_coronal*.png \${image}_axial*.png \
                \${image}_sagittal*.png \${image}_mosaic.png
            convert -annotate +20+230 "\${title}" -fill white -pointsize 30 \
                \${image}_mosaic.png \${image}_mosaic.png
            # Clean up.
            rm \${image}_coronal*.png \${image}_sagittal*.png \${image}_axial*.png
        done
        # Create GIF.
        convert -delay 10 -loop 0 -morph 10 \
            warped_mosaic.png fixedimage_mosaic.png warped_mosaic.png \
            ${prefix}_${suffix_qc}_registration_ants_mqc.gif
        # Clean up.
        rm *_mosaic.png
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9.a-zA-Z-]+).*/\\1/')
        mrtrix: \$(mrinfo -version 2>&1 | grep "== mrinfo" | sed -E 's/== mrinfo ([0-9.]+).*/\\1/')
        imagemagick: \$(convert -version | grep "Version:" | sed -E 's/.*ImageMagick ([0-9.-]+).*/\\1/')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    set +e
    function handle_code () {
    local code=\$?
    ignore=( 1 )
    [[ " \${ignore[@]} " =~ " \$code " ]] || exit \$code
    }
    trap 'handle_code' ERR

    antsRegistrationSyNQuick.sh -h
    antsApplyTransforms -h
    convert -h
    scil_viz_volume_screenshot -h

    touch ${prefix}__t1_warped.nii.gz
    touch ${prefix}__output0GenericAffine.mat
    touch ${prefix}__output1InverseAffine.mat
    touch ${prefix}__output0InverseWarp.nii.gz
    touch ${prefix}__output1Warp.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9.a-zA-Z-]+).*/\\1/')
        mrtrix: \$(mrinfo -version 2>&1 | grep "== mrinfo" | sed -E 's/== mrinfo ([0-9.]+).*/\\1/')
        imagemagick: \$(convert -version | grep "Version:" | sed -E 's/.*ImageMagick ([0-9.-]+).*/\\1/')
    END_VERSIONS
    """
}
