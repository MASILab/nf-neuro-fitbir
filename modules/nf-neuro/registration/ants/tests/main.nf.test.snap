{
    "registration - ants": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,6491f46d2c2cae09338effb989ec3bf7"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0GenericAffine.mat:md5,3b6d0eaddd216d15bf50a116c2f09065"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1InverseAffine.mat:md5,14b90a8217065e42b21102de73a489c1"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1Warp.nii.gz:md5,db8e148adf0e32d551cf5e75ca3e76ab"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0InverseWarp.nii.gz:md5,3069b784eaa199078d21ecbd5366e27b"
                    ]
                ],
                "5": [

                ],
                "6": [
                    "versions.yml:md5,b79ebe5518288b629eb72580d6e6b785"
                ],
                "affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0GenericAffine.mat:md5,3b6d0eaddd216d15bf50a116c2f09065"
                    ]
                ],
                "image": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,6491f46d2c2cae09338effb989ec3bf7"
                    ]
                ],
                "inverse_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1InverseAffine.mat:md5,14b90a8217065e42b21102de73a489c1"
                    ]
                ],
                "inverse_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0InverseWarp.nii.gz:md5,3069b784eaa199078d21ecbd5366e27b"
                    ]
                ],
                "mqc": [

                ],
                "versions": [
                    "versions.yml:md5,b79ebe5518288b629eb72580d6e6b785"
                ],
                "warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1Warp.nii.gz:md5,db8e148adf0e32d551cf5e75ca3e76ab"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T21:59:53.198912314"
    },
    "registration - ants - quick": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,45822afae6c0dc33551b690f6f10dbe4"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0GenericAffine.mat:md5,17e711c6a2f73415269b4272344890bc"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1InverseAffine.mat:md5,e5df1f989e0340746d9742ff2d0a7a99"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1Warp.nii.gz:md5,813a4218d33d5acfebadc4ba0e49df8b"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0InverseWarp.nii.gz:md5,813a4218d33d5acfebadc4ba0e49df8b"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test_T1_to_DWI_registration_ants_mqc.gif:md5,9cdd7e9bf7facc3f95a61de03e15d47d"
                    ]
                ],
                "6": [
                    "versions.yml:md5,b79ebe5518288b629eb72580d6e6b785"
                ],
                "affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0GenericAffine.mat:md5,17e711c6a2f73415269b4272344890bc"
                    ]
                ],
                "image": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,45822afae6c0dc33551b690f6f10dbe4"
                    ]
                ],
                "inverse_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1InverseAffine.mat:md5,e5df1f989e0340746d9742ff2d0a7a99"
                    ]
                ],
                "inverse_warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0InverseWarp.nii.gz:md5,813a4218d33d5acfebadc4ba0e49df8b"
                    ]
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test_T1_to_DWI_registration_ants_mqc.gif:md5,9cdd7e9bf7facc3f95a61de03e15d47d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b79ebe5518288b629eb72580d6e6b785"
                ],
                "warp": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1Warp.nii.gz:md5,813a4218d33d5acfebadc4ba0e49df8b"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T22:00:25.806953541"
    },
    "registration - ants - options": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,c090343bdf2af2eac20a051b3cf9bb20"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0GenericAffine.mat:md5,d3230976041b8532b14a00e2e5760395"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1InverseAffine.mat:md5,10a6c2eededdc85709bbff395f0aab3a"
                    ]
                ],
                "3": [

                ],
                "4": [

                ],
                "5": [
                    [
                        {
                            "id": "test"
                        },
                        "test__registration_ants_mqc.gif:md5,1bfa145273f0ece8a2293efc7d2a8c2e"
                    ]
                ],
                "6": [
                    "versions.yml:md5,b79ebe5518288b629eb72580d6e6b785"
                ],
                "affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output0GenericAffine.mat:md5,d3230976041b8532b14a00e2e5760395"
                    ]
                ],
                "image": [
                    [
                        {
                            "id": "test"
                        },
                        "test__warped.nii.gz:md5,c090343bdf2af2eac20a051b3cf9bb20"
                    ]
                ],
                "inverse_affine": [
                    [
                        {
                            "id": "test"
                        },
                        "test__output1InverseAffine.mat:md5,10a6c2eededdc85709bbff395f0aab3a"
                    ]
                ],
                "inverse_warp": [

                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test__registration_ants_mqc.gif:md5,1bfa145273f0ece8a2293efc7d2a8c2e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b79ebe5518288b629eb72580d6e6b785"
                ],
                "warp": [

                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T22:00:54.46748032"
    },
    "registration - ants - stub": {
        "content": [
            [
                "versions.yml:md5,b79ebe5518288b629eb72580d6e6b785"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T22:01:09.70219899"
    }
}
